sympy/matrices/determinant.py

```diff
@@ -409,10 +415,13 @@ def _charpoly(M, x='lambda', simplify=_simplify):
             m = m * (x - simplify(i))
         return PurePoly(m, x)
 
-    berk_vector = _berkowitz_vector(M)
-    x = uniquely_named_symbol(x, berk_vector, modify=lambda s: '_' + s)
+    types = list(map(type, M))
+    if all(issubclass(t, Expr) for t in types):
+        berk_vector = _berkowitz_vector(M)
+        x = uniquely_named_symbol(x, berk_vector, modify=lambda s: '_' + s)
+        return PurePoly([simplify(a) for a in berk_vector], x)
 
-    return PurePoly([simplify(a) for a in berk_vector], x)
+    return _charpoly_DOM(M, x)
```

Comment:
You should use `set(map(type, M))` because I think it will make the `issubclass` check faster.

It looks like this will not call `_charpoly_DOM` if all elements in the matrix are Expr (which should be the case for any reasonable matrix).

Reply:
> You should use `set(map(type, M))` because I think it will make the `issubclass` check faster.
> 
Sure, will do so.

> It looks like this will not call `_charpoly_DOM` if all elements in the matrix are Expr (which should be the case for any reasonable matrix).
>
But this is what we wanted as of now, right? 

As mentioned above, generally `EX `will not be faster than `Expr` but there will be some cases where `DomainMatrix `with `EX` is faster than `Matrix` with `Expr`.

Otherwise, we can check for domain `EX` specifically and implement `Samuelson-Berkowitz `specifically for it, else we can use `_charpoly_DOM` with `simplify` for `EX`, but this will be a little slower for sure.

Reply:
In all of the cases where we will want to use DomainMatrix the entries of the Matrix will be of type Expr. However the domain used to represent those entries as a DomainMatrix depends on what kind of Expr they are. For example if all entries are Integer (a subclass of Expr) then we can use the ZZ domain. For the ZZ domain DomainMatrix is always faster and we should always use it.